### ISSUE TITLE
[$1000] CRITICAL: An account is mentioned in a new thread but neither see that thread report in LHN nor receive reply notification

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21146]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login with two accounts on two devices (Account A, Account B)
2. Create a group chat including A and B
3. A mentions B in the group chat
4. A clicks Reply in thread that message
5. Verify that B does not see the Thread report in LHN
6. In that thread report, A replies with any message
7. Verify that B does not receive any reply notification
## Expected Result: 
B sees the new thread report in LHN and receives reply notification
## Actual Result:  
B cannot see the new thread report in LHN and receive any reply notification

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.29-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/dbfb2fab-9a9f-4fcc-8ef6-f8bee8c355ea

[Screencast from 12-06-2023 12_02_15.webm](https://github.com/Expensify/App/assets/43996225/4af61318-0a86-4d78-b585-97133fbc06c5)

**Expensify/Expensify Issue URL:**
**Issue reported by:** @tienifr 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686546659179169

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01712438b34b131aed</li>
        <li>Upwork Job ID: 1673396102439284736</li>
        <li>Last Price Increase: 2023-06-26</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @jasperhuangg</details>

**Comment by johncschuster** [comment]
Looks like we're waiting on proposals. Pinging the Open Source channel.

**Comment by bernhardoj** [comment]
I think this is expected because B needs to join the thread before receiving any messages in that thread.

**Comment by hannojg** [comment]
I would say this is most likely a backend issue to which we external contributors have no access.
If it wouldn't be a backend issue the chat would appear e.g. when reloading the app and refetching the reports.

**Comment by johncschuster** [comment]
Thanks for calling that out, @hannojg! I'll switch the label to `Internal`

**Comment by johncschuster** [comment]
Thanks for the callout, @fedirjh!

**Comment by chiragsalian** [comment]
Interesting. This is quite the feature to implement. Its a good feature but currently threads does not do much with mentions. We will be changing our notification strategy for threads so i think we should HOLD this issue until this [one](https://github.com/Expensify/Expensify/issues/295764) is done. There might not be a lot of overlap but i want to avoid duplicate work and the other issue is higher on our priority list. So placing this on HOLD until we get that one done first.

For context https://github.com/Expensify/Expensify/issues/295764, makes the following changes,
1. Viewing a thread alone will not make the user subscribe to it to receive new message.
2. A user will subscribe to a thread only if they change notification preferences or leave a message.

**Comment by johncschuster** [comment]
Thanks, @chiragsalian! Given we're putting this one on hold, I'm going to downgrade it to a `Weekly` cadence.

**Comment by johncschuster** [comment]
Reassigning since Chirag is on leave.

**Comment by johncschuster** [comment]
We're still waiting on the above linked issue.

**Comment by amyevans** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still waiting on the linked issue to be resolved.

**Comment by amyevans** [comment]
Still on hold for https://github.com/Expensify/Expensify/issues/295764 though there are PRs in review. @chiragsalian are you able to take this one on now that you're back?

**Comment by johncschuster** [comment]
Still on hold. @chiragsalian, are you able to take this one now that you're back?

**Comment by johncschuster** [comment]
@chiragsalian bump!

**Comment by chiragsalian** [comment]
Just catching up here, 
1. Firstly this is off HOLD so i updated the title.
2. Shouldn't this issue be predesigned before implementing? I'm asking because I'm not sure if every mention should be automatically be invited to a room and receive notification. For example, what is userA mentioned userB in a private room. I doubt they would want userB to be invited. So maybe we've to discuss how this functionality should work in different scenarios. Perhaps we'll want a whisper like slack to give userA the option to invite userB or do nothing if they are not part of the parent report? Or do we allow this feature to work for now only if userB has access to parent report?

**Comment by chiragsalian** [comment]
As for if i can work on this. My plate is a bit full catching up on stuff. Until i clear up my plate i won't be able to actively work on this. If amy cannot work on it she could unassign it and then if one of us or someone else gets free they can work on this.

**Comment by amyevans** [comment]
Sounds good! I do agree and think this is worthy of a pre-design fwiw

**Comment by johncschuster** [comment]
Thanks @chiragsalian and @amyevans! I'm going to chat with Puneet about this to see if it should be considered in the `@mentions` pre-designs.

**Comment by johncschuster** [comment]
Ah, I forgot to update this. Ok! I chatted with @puneetlath, and it sounds like @jasperhuangg is building `invite-via-mention`. So very soon, when you mention someone in a chat, it'll invite them to it.

**Comment by johncschuster** [comment]
That said, I _believe_ this issue will be resolved with that feature, but I'll let @jasperhuangg confirm.

**Comment by johncschuster** [comment]
Hey @jasperhuangg! Can you chime in when you get a moment? We're trying to figure out if `invite-via-mention` will resolve the behavior we're seeing here. We'd love your take!

**Comment by johncschuster** [comment]
@jasperhuangg bump! What are your thoughts on the above?

**Comment by jasperhuangg** [comment]
Hey guys, apologies for the delay here, my K2 has been extremely overfilled after trying to meet a deadline for Money2020, and I'm now working towards clearing up a lot of my overdue `Daily` issues.

**Comment by jasperhuangg** [comment]
So this is an interesting problem.

I think what's going on is this:
1. userA mentions userB in a comment.
2. userA starts a thread off of this comment. This thread report is created in the back-end, but isn't shared with userB, who was mentioned in its parent report action.
3. When userA comments in the thread, nothing is sent to userB since the thread report isn't shared with them.

cc @chiragsalian since this has to do with threads, do you think my rationale on this seems correct?

I think the path forward here is to share underlying thread reports with anyone mentioned in the parent report action as soon as they're created (or as soon as the first comment is added into the thread), since being mentioned in the parent report action have the same behavior as getting mentioned directly in the thread, correct?

**Comment by jasperhuangg** [comment]
I've opened a thread in Slack [here](https://expensify.slack.com/archives/C03TQ48KC/p1698709550010549) for more discussion

**Comment by jasperhuangg** [comment]
Conversation is happening in the thread. We've concluded that we should be sharing everyone who is mentioned in a message if a thread report is created from that message.

There are a few followup questions that I have, but barring those we should be able to start implementing a solution for this.

**Comment by jasperhuangg** [comment]
Gonna bump the priority to Daily on this one

**Comment by chiragsalian** [comment]
replied in slack 🙂

**Comment by jasperhuangg** [comment]
We've come to a consensus in Slack as to how this should work! Thanks to everyone who got involved in that thread for the great discussion:

#### when a thread report is created, it’s shared with the following people, all with the “hidden” preference:
- the person creating the thread
- the person who created the parent reportAction
- anyone mentioned in the parent reportAction

#### then whenever someone writes a message in that thread
- the preference is updated to “always” for all the people above
- the preference stays “hidden” for anyone else that is shared on the thread

**Comment by jasperhuangg** [comment]
Gonna try to get a PR up for this soon

**Comment by jasperhuangg** [comment]
PR up! Been working on testing it E2E, still need to write some automated tests for it.

**Comment by jasperhuangg** [comment]
PR is tested and ready for review

**Comment by quinthar** [comment]
Are these dupes? https://github.com/Expensify/App/issues/30919

**Comment by jasperhuangg** [comment]
@johncschuster this was merged and deployed. I believe we can sort out payment for this issue and close it out?-

**Comment by johncschuster** [comment]
Sorry I missed your comment, @jasperhuangg! I'll get payment issued.

**Comment by johncschuster** [comment]
@jasperhuangg, I've just reviewed the linked PRs above and I'm having a hard time tracking down who needs to be paid for this specific issue.

Given the date when this issue was reported, it looks like we will need to issue a Bug Report payment to @tienifr. I see @fedirjh is assigned to the issue as well, but I don't see anyone assigned to the issue.

@fedirjh, did you review any PRs for this issue? If so, can you point them out for me? Thanks!

**Comment by johncschuster** [comment]
Just a heads up, I'm OOO today through January 2, but I will get payment issued as soon as I'm back!

**Comment by jasperhuangg** [comment]
Thank you @fedirjh I agree with what he's said.

**Comment by jasperhuangg** [comment]
bump @johncschuster! We clarified how payment needs to happen [here](https://github.com/Expensify/App/issues/21146#issuecomment-1861070247).

**Comment by johncschuster** [comment]
Thanks for the bump! I'll get this sorted today!

**Comment by johncschuster** [comment]
Sorry for the delay, folks! I had to reach back into the archives to figure out the right payment amount for the bug report 😅 

@tienifr, I've just sent you an [offer](https://www.upwork.com/ab/applicants/1755605452030951424/job-details) of $250 for the bug report. Can you accept that so I can issue payment? Thank you!

**Comment by johncschuster** [comment]
Paid! Thanks, @tienifr!

Looks like all we need now is the regression test steps. @fedirjh, will you be taking care of that?

**Comment by jasperhuangg** [comment]
Sorry for the delay, it's been a while since I merged that PR 😆 

You need to be signed in simultaneously to all 4 accounts.
1. Navigate to a group chat with 4 people total. As userA, make a comment that mentions userB and userC.
2. Create a thread from this comment.
3. In all the other accounts, verify the thread report exists in Onyx.
4. Now comment in the thread as userD.
5. Verify users A, B and C receive the comment and that the thread report appears in their LHNs.

👍 or 👎
